python/minetester/minetest_env.py

```diff
@@ -0,0 +1,505 @@
+
+        # TODO more robust check for whether a server/client is alive while receiving observations
+        if self.client_process is not None and self.client_process.poll() is not None:
+            return self.last_obs, 0.0, True, False, {}
+
+        # Receive observation
+        logging.debug("Waiting for obs...")
+        byte_obs = self.socket.recv()
+        next_obs, rew, done = unpack_pb_obs(byte_obs)
+
+        self.last_obs = next_obs
+        logging.debug(f"Received obs - {next_obs.shape}; reward - {rew}")
+
+        if self.render_mode == "human":
+            self._display_pygame()
+
+        return next_obs, rew, done, False, {}
+
+    def render(self):
+        if self.render_mode == "human":
+            # rendering happens during step, as per gymnasium API
+            return None
+        return self.last_obs
+
+    def close(self):
+        if self.socket is not None:
+            self.socket.close()
+        # TODO improve process termination
+        # i.e. don't kill, but close signal
+        if self.client_process is not None:
+            self.client_process.kill()
+        if self.reset_world:
+            self._delete_world()
+        if self.clean_config:
+            self._delete_config()
+
+    def _start_minetest_client(
+        self,
+        log_path: str,
+    ):
+        cmd = [
+            self.minetest_executable,
+            "--go",  # skip menu
+            "--config",
+            self.config_path,
+            "--remote-input",
+            f"127.0.0.1:{self.env_port}",
+        ]
+        if self.headless:
+            cmd.append("--headless")
+        if self.verbose_logging:
+            cmd.append("--verbose")
+        if self.world_dir is not None:
+            cmd.extend(["--world", self.world_dir])
+        if self.gameid is not None:
+            cmd.extend(["--gameid", self.gameid])
+
+        stdout_file = log_path.format("client_stdout")
+        stderr_file = log_path.format("client_stderr")
+        with open(stdout_file, "w") as out, open(stderr_file, "w") as err:
+            client_env = os.environ.copy()
+            if self.x_display is not None:
+                client_env["DISPLAY"] = ":" + str(self.x_display)
+            # enable GPU usage
+            client_env["__GLX_VENDOR_LIBRARY_NAME"] = "nvidia"
+            client_env["__NV_PRIME_RENDER_OFFLOAD"] = "1"
+            # disable vsync
+            client_env["__GL_SYNC_TO_VBLANK"] = "0"
+            client_env["vblank_mode"] = "0"
+            out.write(
+                f"Starting client with command: {' '.join(str(x) for x in cmd)}\n"
+            )
+            out.write(f"Client environment: {client_env}\n")
+            client_process = subprocess.Popen(
+                cmd, stdout=out, stderr=err, env=client_env
+            )
+            out.write(f"Client started with pid {client_process.pid}\n")
+        return client_process
+
+
+def unpack_pb_obs(received_obs: str):
```

Comment:
I'd name this `deserialize_obs`
deserialize is a more specific term than unpack